### ISSUE TITLE
Unable to pull docker images from GitHub Docker registry in RunnerDeployment resource on Fargate

### ISSUE THREAD

**Issue by nyagnik** [actions-runner-controller/actions-runner-controller#637]
I have built custom actions-runner image (basically installed a couple of packages on top of actions runner) as described here: https://github.com/actions-runner-controller/actions-runner-controller#software-installed-in-the-runner-image
The image is in GitHub Docker private registry:
```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: my-runner
spec:
  template:
    spec:
      organization: my-org
      group: my-org-group
      image: docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
      imagePullSecrets:
      - name: gh-docker-registry
      dockerEnabled: false
      resources:
        limits:
          cpu: "3.0"
          memory: "3Gi"
        requests:
          cpu: "2.0"
          memory: "1Gi"
      labels:
        - my-runner
  ```
This always fails with pulling the image.
```
$ kubectl describe pod my-runner-pod
Events:
  Type     Reason           Age                     From               Message
  ----     ------           ----                    ----               -------
  Warning  Failed           <invalid> (x3 over 8s)  kubelet            Failed to pull image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner": rpc error: code = NotFound desc = failed to pull and unpack image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner": failed to copy: httpReaderSeeker: failed open: content at https://docker.pkg.github.com/v2/my-org/runner-image/custom-summerwind-runner/manifests/sha256:f98bbec75f1d24b97b5e172e891fee5d82195641e92f5b0a21e67675d399c372 not found: not found
  Warning  Failed           <invalid> (x3 over 8s)  kubelet            Error: ErrImagePull
```
When I push the same image in Docker Hub or when I use the same image from GH as normal Kubernetes deployment, the cluster is able to pull the image.

**Comment by frbk** [comment]
Can you run `kubectl describe pod my-runner-pod` and check which image it uses. I think I had a similar issue where my image looked like this: `customrepo/customimage:latest:0.19v`. Not sure why, but in my case `:0.19v` was getting appending to my image. Describe command will show which image is used if you scroll up the output a bit. I was planning on opening an issue but you beat me to it.

**Comment by nyagnik** [comment]
@frbk for me, no, it did not add any additional tag to my image.
```
$ kubectl describe pod my-runner-pod-6s8qs-h2lgn
<output-scrapped>
Containers:
  runner:
    Container ID:   
    Image:          docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
    Image ID:       
    Port:           <none>
    Host Port:      <none>
    State:          Waiting
      Reason:       ImagePullBackOff
    Ready:          False
<output-scrapped>
Events:
  Type     Reason           Age                    From               Message
  ----     ------           ----                   ----               -------
  Normal   Scheduled        6m17s                  fargate-scheduler  Successfully assigned default/my-runner-pod-6s8qs-h2lgn to fargate-ip-10-332-91-118.ec2.internal
  Normal   Pulling          4m57s (x4 over 6m17s)  kubelet            Pulling image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner"
  Warning  Failed           4m56s (x4 over 6m16s)  kubelet            Failed to pull image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner": rpc error: code = NotFound desc = failed to pull and unpack image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner:latest": failed to copy: httpReaderSeeker: failed open: content at https://docker.pkg.github.com/v2/my-org/runner-image/custom-summerwind-runner/manifes
ts/sha256:f98bbec75f1d24b97b5e172e891fee5d82195641e92f5b0a21e67675d399c372 not found: not found
  Warning  Failed           4m56s (x4 over 6m16s)  kubelet            Error: ErrImagePull
  Normal   BackOff          4m19s (x7 over 6m16s)  kubelet            Back-off pulling image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner"
  Warning  Failed           71s (x21 over 6m16s)   kubelet            Error: ImagePullBackOff
```
CC: @mumoshu @callum-tait-pbx

**Comment by frbk** [comment]
@nyagnik Are you using a private docker repo? Have you created a secret with credentials? Have a look at this: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line

**Comment by nyagnik** [comment]
@frbk yes, I am pulling from private Github docker repo, I have also referenced the `imagePullSecrets` in my `runnerDeployment` resource.
```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: my-runner
spec:
  template:
    spec:
      organization: my-org
      group: my-org-group
      image: docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
      imagePullSecrets:
      - name: gh-docker-registry
      dockerEnabled: false
      resources:
        limits:
          cpu: "3.0"
          memory: "3Gi"
        requests:
          cpu: "2.0"
          memory: "1Gi"
      labels:
        - my-runner
```
This issue is happening only when I use private Github docker registry repo, it works fine when I use private Docker hub registry.

**Comment by frbk** [comment]
@nyagnik I would probably look into how you are authenticating with github if docker hub works. Are you using 2fa with github? Does your token have correct permissions?

**Comment by mumoshu** [comment]
Hey! RunnerDeployment uses K8s "Pod" to let Kubernetes pull images for you. So this should have nothing to do with actions-runner-controller.

Generally speaking, please ensure that your config allows any regular K8s pod to pull images from your registry. And then try it with RunnerDeployment.

From what I can see from your log, this might be due to either (1) your pull secret isn't valid or (2)you haven't pushed your images to the registry yet.

I don't even know if you understand the basic usage of ghcr on Kubernetes in general, please forgive me just recommending you to read something like https://dev.to/asizikov/using-github-container-registry-with-kubernetes-38fb so that you can start by setting it up correctly for pods.

**Comment by nyagnik** [comment]
@frbk @mumoshu I am able to successfully pull and run the same image via `Deployment` resource without any issues.
the issue is occurring only when I try the same via `RunnerDeployment` resource.

Reference:
```
$ cat test-deploy.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: test
  name: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: test
    spec:
      containers:
      - image: docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
        name: custom-summerwind-runner
        resources: {}
      imagePullSecrets:
      - name: gh-docker-registry 
status: {}
```
```
$ kubectl create -f test-deploy.yaml
$ kubectl describe pod test-6db8dc6957-qc4ft
  Normal   Pulling    2m37s (x5 over 4m16s)  kubelet            Pulling image "docker.pkg.github.com/eaton-vance-corp/terraform-eks-runner-module/jonico-summerwind-runner"
  Normal   Pulled     2m37s                  kubelet            Successfully pulled image "docker.pkg.github.com/eaton-vance-corp/terraform-eks-runner-module/jonico-summerwin
d-runner" in 210.832423ms
```
Hence this proves the pull secret is valid, and indeed the image is exist. 
I am getting the issue only when I try to deploy the same with `RunnerDeployment` as mentioned in my previous comments.

**Comment by mumoshu** [comment]
@nyagnik Thanks a lot! That helps investigating this.

How does your `Pod` created by the `RunnerDeployment` looks like? Does it have the equivalent `imagePullSecrets` field in it, have a wrong one, or totally miss it? There should be a pod whose name is prefixed with your runnerdeployment's name, `my-runner`.

**Comment by nyagnik** [comment]
yes, it has the same equivalent `imagePullSecrets.
This is my `RunnerDeployment` manifest:
```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: my-runner
spec:
  template:
    spec:
      organization: my-org
      group: my-org-group
      image: docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
      imagePullSecrets:
      - name: gh-docker-registry
      dockerEnabled: false
      resources:
        limits:
          cpu: "3.0"
          memory: "3Gi"
        requests:
          cpu: "2.0"
          memory: "1Gi"
      labels:
        - my-runner
```
My pod manifest:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    CapacityProvisioned: 2vCPU 4GB
    Logging: 'LoggingDisabled: LOGGING_CONFIGMAP_NOT_FOUND'
    kubernetes.io/psp: eks.privileged
  creationTimestamp: "2021-06-21T06:19:24Z"
  labels:
    eks.amazonaws.com/fargate-profile: runner_fargate_profile
    pod-template-hash: 749fd4569f
    runner-deployment-name: my-runner
    runner-template-hash: 68965594cb
  name: my-runner-6s8qs-h2lgn
  namespace: default
  ownerReferences:
  - apiVersion: actions.summerwind.dev/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Runner
    name: my-runner-6s8qs-h2lgn
    uid: e0767d78-c16a-4b2f-b6a5-734514fa0f6a
  resourceVersion: "5309959"
  selfLink: /api/v1/namespaces/default/pods/my-runner-6s8qs-h2lgn
  uid: 8b4f3767-1234-48bb-8704-3a14cd3058d6
spec:
  containers:
  - env:
    - name: RUNNER_NAME
      value: my-runner-pod-6s8qs-h2lgn
    - name: RUNNER_ORG
      value: Eaton-Vance-Corp
    - name: RUNNER_REPO
    - name: RUNNER_ENTERPRISE
    - name: RUNNER_LABELS
      value: my-runner
    - name: RUNNER_GROUP
      value: my-org-group
    - name: RUNNER_TOKEN
      value: <output-scrapped>
    - name: DOCKERD_IN_RUNNER
      value: "false"
    - name: GITHUB_URL
      value: https://github.com/
    - name: RUNNER_WORKDIR
      value: /runner/_work
    image: docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
    imagePullPolicy: Always
    name: runner
    resources:
      limits:
        cpu: "3"
        memory: 3Gi
      requests:
        cpu: "2"
        memory: 1Gi
    securityContext: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /runner
      name: runner
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-cn57l
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: gh-docker-registry
  nodeName: fargate-ip-10-234-91-92.ec2.internal
  preemptionPolicy: PreemptLowerPriority
  priority: 2000001000
  priorityClassName: system-node-critical
  restartPolicy: OnFailure
  schedulerName: fargate-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir: {}
    name: runner
  - name: default-token-cn57l
    secret:
      defaultMode: 420
      secretName: default-token-cn57l
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:20:34Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:20:34Z"
    message: 'containers with unready status: [runner]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:20:34Z"
    message: 'containers with unready status: [runner]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:20:34Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner
    imageID: ""
    lastState: {}
    name: runner
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        message: Back-off pulling image "docker.pkg.github.com/my-org/runner-image/custom-summerwind-runner"
        reason: ImagePullBackOff
  hostIP: 10.223.91.92
  phase: Pending
  podIP: 10.223.91.92
  podIPs:
  - ip: 10.223.91.92
  qosClass: Burstable
  startTime: "2021-06-21T06:20:34Z"
  ```

**Comment by mumoshu** [comment]
@nyagnik Thanks! The pod spec seems good so probably this isn't actions-runner-controller's issue.

Ah, Fargate!! Unfortunately, we don't have support for Fargate. The only user I know of who has successfully ran actions-runner-controller and runner pods on Fargate is https://github.com/actions-runner-controller/actions-runner-controller/issues/631 and he doesn't seem to use any private registry to me. So you might be the first one.

Probably you'd better ask AWS support like "I'm seeing Fargate unable to pull this container image using this pod spec containing a valid image pull secret".

My quick googling revealed https://github.com/aws/containers-roadmap/issues/1012, which seems relevant to your issue.

**Comment by nyagnik** [comment]
@mumoshu You're right, the issue is with Fargate.
I have reached out to AWS support, they've mentioned Fargate follows contained standard for the images which is something different from Docker standard.
There are two URLs in GitHub:
1. docker.pkg.github.com - Follows docker standard
2. ghcr.io - Follows contained standard.

So I had to use ghcr.io for hosting my images, finally, it worked.

Thanks @mumoshu @frbk :).
